crates/matrix-sdk-crypto/src/verification/requests.rs

```diff
@@ -663,16 +703,48 @@ impl VerificationRequest {
     ) -> Result<(), CryptoStoreError> {
         let inner = self.inner.get();
 
-        let InnerRequest::Ready(s) = inner else {
-            warn!(
-                sender = sender.as_str(),
-                device_id = content.from_device().as_str(),
-                "Received a key verification start event but we're not yet in the ready state"
-            );
-            return Ok(());
-        };
+        match &inner {
+            InnerRequest::Created(_)
+            | InnerRequest::Requested(_)
+            | InnerRequest::Passive(_)
+            | InnerRequest::Done(_)
+            | InnerRequest::Cancelled(_) => {
+                warn!(
+                    sender = sender.as_str(),
+                    device_id = content.from_device().as_str(),
+                    "Received a key verification start event but we're not yet in the ready state"
+                );
+                Ok(())
+            }
+            InnerRequest::Ready(s) => {
+                let s = s.clone();
+                drop(inner);
+
+                if let Some(new_state) = s
+                    .receive_start(sender, content, self.we_started, self.inner.clone().into())
+                    .await?
+                {
+                    let mut inner = self.inner.write();
+                    ObservableWriteGuard::set(&mut inner, InnerRequest::Transitioned(new_state));
+                }
 
-        s.receive_start(sender, content, self.we_started, self.inner.clone().into()).await
+                Ok(())
+            }
+            InnerRequest::Transitioned(s) => {
+                let s = s.clone();
+                drop(inner);
+
+                if let Some(new_state) = s
+                    .receive_start(sender, content, self.we_started, self.inner.clone().into())
+                    .await?
+                {
+                    let mut inner = self.inner.write();
+                    ObservableWriteGuard::set(&mut inner, InnerRequest::Transitioned(new_state));
+                }
+
+                Ok(())
+            }
```

Comment:
Is this branch expected? The branches for any other state than Ready seem to indicate this should be in the `Ready` state, and this branch's body seems copy/pasted from the `Ready` branch. If that's expected, can you add a comment explaining why it's fine, please?

Reply:
Yeah, it's meant to be for the case where you try to use QR code verification then change your mind. Will add a comment.

Reply:
Added a comment now.